### PR TITLE
Implement navigation for XY charts

### PR DESCRIPTION
**Pull request by IbrahimFradj**
**Navigation implemented :** 

- **Zoom-in by following the mouse cursor :** "w","W" and "ctrl+mouse wheel up".
- **Zoom-out by following the mouse cursor :** "s", "S", and ""ctrl+mouse wheel down".
- **Zooming with range :** click right and drag.
-  **Panning left :** "a","A", "arrowLeft" and "shift+mouse wheel up".
-  **Panning right :**  "d","D", "arrowRight" and "shift+mouse wheel down".

Fixes theia-ide#389

**Here is the result**
zoom

Signed-off-by: Ibrahim Fradj <ibrahim.fradj@ericsson.com>

### CODE REVIEW

File: packages/react-components/src/components/xy-output-component.tsx [hunk 1]
```diff
@@ -284,10 +287,128 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
     }
 
+    private zoomIn(diffStartMouse: number, diffEndMouse: number, percentRange: number) {
+        if (this.isRangeMove) {
+            this.props.unitController.viewRange = {
+                start: this.props.unitController.viewRange.start + percentRange,
+                end: this.props.unitController.viewRange.end - percentRange
+            };
+        } else {
+            const newStartRange = this.xStartPos - (diffStartMouse - (diffStartMouse * 0.1));
```
Comment (MatthewKhouzam) [inline packages/react-components/src/components/xy-output-component.tsx:302]: extract .1 as a constant

File: packages/react-components/src/components/xy-output-component.tsx [hunk 2]
```diff
@@ -284,10 +287,128 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+        const diffEndMouse = this.props.unitController.viewRange.end - this.xStartPos;
+        const percentRange = this.props.unitController.viewRangeLength * 0.1;
+        if (this.xStartPos >= this.props.unitController.viewRange.end || this.xStartPos <= this.props.unitController.viewRange.start) {
+            this.isRangeMove = true;
+        } else {
+            this.isRangeMove = false;
+        }
+        switch (key.key) {
+            case 'W':
+            case '+':
```
Comment (MatthewKhouzam) [inline packages/react-components/src/components/xy-output-component.tsx:388]: please add '=' as it is no shift '+'

File: packages/react-components/src/components/xy-output-component.tsx [hunk 3]
```diff
@@ -284,10 +287,128 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+            case 'w': {
+                this.zoomIn(diffStartMouse, diffEndMouse, percentRange);
+                break;
+            }
+            case 'S':
+            case '-':
```
Comment (MatthewKhouzam) [inline packages/react-components/src/components/xy-output-component.tsx:394]: add '_'

File: packages/react-components/src/components/xy-output-component.tsx [hunk 4]
```diff
@@ -284,10 +287,128 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+            case 's': {
+                this.zoomOut(diffStartMouse, diffEndMouse, percentRange);
+                break;
+            }
+            case 'A':
```
Comment (MatthewKhouzam) [inline packages/react-components/src/components/xy-output-component.tsx:399]: question... not a suggestion: should we add right arrow and left arrow too?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 5]
```diff
@@ -284,10 +287,128 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+                start: Math.max(0, newStartRange),
+                end: Math.min(newEndRange, this.props.unitController.absoluteRange)
+            };
+        }
+    }
+
+    private shiftLeft(percentRange: number) {
+        let startRange = this.props.unitController.viewRange.start - percentRange;
+        let endRange = this.props.unitController.viewRange.end - percentRange;
+        if (startRange <= 0) {
```
Comment (MatthewKhouzam) [inline packages/react-components/src/components/xy-output-component.tsx:331]: make a helper function

File: packages/react-components/src/components/xy-output-component.tsx [hunk 6]
```diff
@@ -284,10 +287,128 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+                end: endRange
+            };
+        }
+
+    }
+
+    private shiftRight(percentRange: number) {
+        let startRange = this.props.unitController.viewRange.start + percentRange;
+        let endRange = this.props.unitController.viewRange.end + percentRange;
+        if (endRange >= this.props.unitController.absoluteRange) {
```
Comment (MatthewKhouzam) [inline packages/react-components/src/components/xy-output-component.tsx:350]: use the helper function here

**Review by MatthewKhouzam — COMMENTED** [review]
Commenting, these are requested changes but I don't want to slow you down.

File: packages/react-components/src/components/xy-output-component.tsx (hunk 2 above)
Comment (MatthewKhouzam) [reply packages/react-components/src/components/xy-output-component.tsx:388]: Holding down on 'W' or 'w' should zoom on the cursor, + on the center.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 7]
```diff
@@ -284,10 +287,128 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+    private zoomOut(diffStartMouse: number, diffEndMouse: number, percentRange: number) {
+        if (this.isRangeMove) {
+            this.props.unitController.viewRange = {
+                start: this.props.unitController.viewRange.start - percentRange,
+                end: this.props.unitController.viewRange.end + percentRange
+            };
+        }
+        else {
+            const newStartRange = this.xStartPos - (diffStartMouse + (diffStartMouse * 0.1));
+            const newEndRange = this.xStartPos + (diffEndMouse + (diffEndMouse * 0.1));
```
Comment (muddana-satish) [inline packages/react-components/src/components/xy-output-component.tsx:320]: Does xy chart zooming and timegraph chart zooming behave the same way? I think the zoom factor used is different. 
Refer to this: https://github.com/theia-ide/timeline-chart/blob/2d3f4d88d8f5f214b5b76136ccdfa25d2d645297/timeline-chart/src/layer/time-graph-chart.ts#L81

Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx]: Should I use the same zoom factor? When I tried to zoom in and out in the timeline view, the zoom in and out is sometimes larger than I want.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 8]
```diff
@@ -61,7 +63,7 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
             collapsedNodes: [],
             orderedNodes: [],
             xyData: {},
-            columns: [{title: 'Name', sortable: true}]
+            columns: [{ title: 'Name', sortable: true }]
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:66]: There seem to be a lot of lines in the diff that are whitespace changes unrelated to adding navigation to XY charts. Unrelated code cleanup changes should at least be in a separate commit (and probably in a different PR). Including unrelated changes adds a lot of noise to the diff and makes the code more difficult to review.

When creating commits with your changes, I recommend reading over the diff and manually staging functional changes using a GUI git tool:
* If the internal git tool in your IDE doesn't allow staging diff sections or individual lines, you can use gitk to create commits.
* First, create a commit with the functional changes. 
* Then, any whitespace or style changes leftover can be committed separately. One way to create a separate PR for leftover style changes would be to "git stash" them, create a new branch from master, then reapply the changes (I think it's "git stash pop").

File: packages/react-components/src/components/xy-output-component.tsx [hunk 9]
```diff
@@ -171,14 +176,14 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
-                {(
+                    {(
                         <i
                             className='fa fa-refresh fa-spin'
                             style={{ marginRight: '5px' }}
                         />
                     )}
                     {
-                    'Analysis running'
+                        'Analysis running'
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:186]: Unrelated whitespace change?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 10]
```diff
@@ -171,14 +176,14 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
                     </Line>
                 </div> :
                 <div className='analysis-running'>
-                {(
+                    {(
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:179]: Unrelated whitespace change?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 11]
```diff
@@ -252,7 +257,7 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
                 newList = newList.concat(id);
             }
         });
-        this.setState({checkedSeries: newList});
+        this.setState({ checkedSeries: newList });
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:260]: Unrelated whitespace change?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 12]
```diff
@@ -266,20 +271,17 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
             newList = newList.concat(id);
         }
         const orderedIds = getAllExpandedNodeIds(nodes, newList);
-        this.setState({collapsedNodes: newList, orderedNodes: orderedIds});
+        this.setState({ collapsedNodes: newList, orderedNodes: orderedIds });
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:274]: Unrelated whitespace change?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 13]
```diff
@@ -266,20 +271,17 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
     }
 
     private onOrderChange(ids: number[]) {
-        this.setState({orderedNodes: ids});
+        this.setState({ orderedNodes: ids });
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:278]: Unrelated whitespace change?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 14]
```diff
@@ -156,12 +158,15 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
                 yAxes: [{ display: false }]
             },
             animation: { duration: 0 },
-            events: [ 'mousedown' ],
+            events: ['mousedown'],
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:161]: Unrelated whitespace change?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 15]
```diff
@@ -82,12 +84,12 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
                     headers.forEach(header => {
-                        columns.push({title: header.name, sortable: true, tooltip: header.tooltip});
+                        columns.push({ title: header.name, sortable: true, tooltip: header.tooltip });
                     });
                 } else {
-                    columns.push({title: 'Name', sortable: true});
+                    columns.push({ title: 'Name', sortable: true });
                 }
-                columns.push({title: 'Legend', sortable: false});
+                columns.push({ title: 'Legend', sortable: false });
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:92]: Unrelated whitespace change?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 16]
```diff
@@ -82,12 +84,12 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
                 const columns = [];
                 if (headers && headers.length > 0) {
                     headers.forEach(header => {
-                        columns.push({title: header.name, sortable: true, tooltip: header.tooltip});
+                        columns.push({ title: header.name, sortable: true, tooltip: header.tooltip });
                     });
                 } else {
-                    columns.push({title: 'Name', sortable: true});
+                    columns.push({ title: 'Name', sortable: true });
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:90]: Unrelated whitespace change?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 17]
```diff
@@ -82,12 +84,12 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
                 const columns = [];
                 if (headers && headers.length > 0) {
                     headers.forEach(header => {
-                        columns.push({title: header.name, sortable: true, tooltip: header.tooltip});
+                        columns.push({ title: header.name, sortable: true, tooltip: header.tooltip });
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:87]: Unrelated whitespace change?

**Review by ebugden — COMMENTED** [review]
I have a lot of feelings about the amount of unrelated whitespace changes mixed in with the commit 😅 I think this is worth correcting before merging for learning purposes (creating clean and specific commits) and also to set a good style example with the commits that are merged.

That being said, the code seems to do what it claims to do! Thanks @IbrahimFradj ☀️

File: packages/react-components/src/components/xy-output-component.tsx (hunk 7 above)
Comment (muddana-satish) [reply packages/react-components/src/components/xy-output-component.tsx]: Okay. I am not sure about that. It would be nice to keep the zooming consistent across the charts. I think we can create a new issue for that.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 18]
```diff
@@ -33,9 +33,9 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
     private lineChartRef: any;
     private mouseIsDown = false;
     private positionZoomMouse = 0;
-    private positionZommkeyboard = 0;
+    private positionZomCenter = 0;
```
Comment (ebugden) [inline packages/react-components/src/components/xy-output-component.tsx:36]: Zom? "positionZomCenter"

Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx:36]: this is the position from where I want to zoom . Because when i zoom with "+" is not the same position when i zoom with "w".
The zoom with "+" follow the center of the view and the zoom with " w" it follows the mouse cursor

Comment (ebugden) [reply packages/react-components/src/components/xy-output-component.tsx:36]: I mean why is it "Zom" rather than "Zoom"?

Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx:36]: ah sorry thank you i didn't notice

File: packages/react-components/src/components/xy-output-component.tsx [hunk 19]
```diff
@@ -61,11 +83,16 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+            columns: [{title: 'Name', sortable: true}],
+            zoomSelectionRangeStart: 0,
+            zoomSelectionRangeEnd: 0,
         };
 
         this.afterChartDraw = this.afterChartDraw.bind(this);
         this.lineChartRef = React.createRef();
+        this.instanceId = XYOutputComponent.instanceCounter++;
+        signalManager().on(Signals.TIMEGRAPH_ZOOMED, this.onTimeGraphZoomed);
+        signalManager().on(Signals.TIMEGRAPH_RESET, this.onTimeGraphReset);
```
Comment (muddana-satish) [inline packages/react-components/src/components/xy-output-component.tsx:95]: Can we listen to these signals in the parent class?

Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx:95]: I tried and I could not put the signals in AbstractTreeOutputComponent

File: packages/react-components/src/components/xy-output-component.tsx [hunk 20]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+    private getPosition(event: React.MouseEvent): number {
+        const offset = this.props.viewRange.getOffset() ?? 0;
+        const scale = this.props.viewRange.getEnd() - this.props.viewRange.getstart();
+        const xPos = this.props.viewRange.getstart() - offset +
+            (event.nativeEvent.offsetX / this.lineChartRef.current.chartInstance.width) * scale;
+        return xPos;
+    }
+
+    private MoveEvent(event: React.MouseEvent) {
+        if (!this.mouseIsDown) {
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx]: Should update the mouse position unconditionally, otherwise for example if you select a range with mouse then W/S, it zooms on the time of the selection start.

I would also only store the mouse x-coordinate, and only calculate the corresponding time when it's used to zoom, that way you don't have to keep updating the mouse 'time' after panning etc..

File: packages/react-components/src/components/xy-output-component.tsx [hunk 21]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+        const diffStartMouse = this.positionZoomMouse - this.props.unitController.viewRange.start;
+        const percentMousePosition = diffStartMouse / this.props.unitController.viewRangeLength;
+        return percentMousePosition;
+    }
+
+    private zoomKey(key: React.KeyboardEvent) {
+        switch (key.key) {
+            case '+':
+            case '=': {
+                this.isZoomCenter = true;
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx]: Not a good idea to set an instance variable so that it can be used by a method you call, it should just be a method parameter.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 22]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+                if (this.props.unitController.viewRangeLength >= 1) {
+                    this.zoomIn(this.initializePosition()[0], this.initializePosition()[1], this.mousePositionPercent());
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx]: These parameters could be calculated from within the called method?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 23]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+        let position = 0;
+        if (this.isZoomCenter) {
+            this.positionZoomCenter = (this.props.unitController.viewRange.end + this.props.unitController.viewRange.start) / 2;
+            position = this.positionZoomCenter;
+        } else {
+            position = this.positionZoomMouse;
+        }
+        const startPosition = position - this.props.unitController.viewRange.start;
+        const endPosition = this.props.unitController.viewRange.end - position;
+        return [startPosition, endPosition];
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:568]: Returns an array that is [duration between window start and mouse, duration between mouse and window end]. That is a very strange API, I would suspect another way would be more understandable...

Also maybe would be cleaner to pass objects instead of array of values.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 24]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+                start: this.NewstartEndRangePanning(right, percentRange)[0],
+                end: this.NewstartEndRangePanning(right, percentRange)[1]
+            };
+        }
+    }
+
+    private shiftLeft() {
+        const numberLeft = -1;
+        const distance = ( this.props.unitController.viewRange.start);
+        this.shift(numberLeft, distance);
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:479]: Would be cleaner to pass a boolean and these two values could be calculated within the called method instead?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 25]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+    }
+
+    private shiftRight() {
+        const numberRight = 1;
+        const distance = (this.props.unitController.absoluteRange - this.props.unitController.viewRange.end);
+        this.shift(numberRight, distance);
+    }
+
+    private wheelEvent(wheel: React.WheelEvent) {
+        if (wheel.shiftKey) {
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:490]: We probably want to do this if only Shift is pressed, mouse wheel with Ctrl+Shift down has weird effect.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 26]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+    }
+
+    private updatePositionAfterPann(right: number, distance: number){
+        this.positionZoomCenter = this.positionZoomCenter + right * distance;
+        this.positionZoomMouse = this.positionZoomMouse + right * distance;
+    }
+
+    private shift(right: number, distance: number) {
+        const percentRange = this.props.unitController.viewRangeLength * TEN_PERCENT;
+        if (this.NewstartEndRangePanning(right, percentRange)[0] < 0) {
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:455]: I would just calculate the expected new range once and handle the limits, not call the same method repeatedly.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 27]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+            case '+':
+            case '=': {
+                this.isZoomCenter = true;
+                if (this.props.unitController.viewRangeLength >= 1) {
+                    this.zoomIn(this.initializePosition()[0], this.initializePosition()[1], this.mousePositionPercent());
+                }
+                break;
+            }
+            case 'W':
+            case 'w': {
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx]: I wonder if we support arrow left/right to pan, if we should also support arrow up/down to zoom?

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/xy-output-component.tsx (hunk 27 above)
Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx]: I asked this question but the answer to this question is that we left the up and down arrows to scroll the page.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 28]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+        }
+        this.updateRange(startPosition, endPosition, percentMousePosition);
+    }
+
+    private zoomOut(startPosition: number, endPosition: number, percentMousePosition: number) {
+        const zoom = -1;
+        this.updateRange( this.newRangeZoom(zoom, startPosition, endPosition)[0], this.newRangeZoom(zoom, startPosition, endPosition)[1],percentMousePosition);
+    }
+
+    private NewstartEndRangePanning(shift: number, percentRange: number): number[] {
```
[the remark this answers is not in the document]
Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx]: yes thank you

File: packages/react-components/src/components/xy-output-component.tsx [hunk 29]
```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+            if (wheel.deltaY < 0) {
+                this.shiftLeft();
+            }
+            else if (wheel.deltaY > 0) {
+                this.shiftRight();
+            }
+        }
+    }
+
+    private getPosition(event: React.MouseEvent): number {
```
[the remark this answers is not in the document]
Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx]: i changed it with getTimeX is it okay ?

File: packages/react-components/src/components/xy-output-component.tsx (hunk 20 above)
Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx]: yes I made these changes in another commit "Zoom in and out by click right drag "when I had this problem.

File: packages/react-components/src/components/xy-output-component.tsx (hunk 22 above)
Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx]: yes

File: packages/react-components/src/components/xy-output-component.tsx (hunk 25 above)
Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx:492]: i didn't have a weird effect when i tried this. it took always the shift

**Comment by mirsky-work** [comment]
Thanks for this nice improvement!
I did notice some misalignment issues between the timeline and XY charts. For example when panning the timeline chart at a high zoom level using 'a' or 'd' shortcuts, the selection range on the XY chart dances around and sometimes ends up misaligned with the selection range on the timeline chart. There are no issues the other way around, and the way to correct the misalignment is to pan the XY chart. See the gif attached.

xy-chart-nav

**Comment by PatrickTasse** [comment]
> I did notice some misalignment issues between the timeline and XY charts. For example when panning the timeline chart at a high zoom level using 'a' or 'd' shortcuts, the selection range on the XY chart dances around and sometimes ends up misaligned with the selection range on the timeline chart.

When the view range is changed rapidly in succession, each small step triggers fetching a new model from the trace-server over the network. These requests are all asynchronous and processed in separate concurrent threads in the trace-server, and the responses do not necessarily come back in the same order that the requests were sent. So if that last response received is not for the last view range that was requested, it can cause a misalignment. This is a design flaw that should be addressed, probably in a separate issue.

**Comment by IbrahimFradj** [comment]
> I did notice some misalignment issues between the timeline and XY charts. For example when panning the timeline chart at a high zoom level using 'a' or 'd' shortcuts, the selection range on the XY chart dances around and sometimes ends up misaligned with the selection range on the timeline chart. There are no issues the other way around, and the way to correct the misalignment is to pan the XY chart. See the gif attached.

I tried what you did in the master branch and found that there was the same issue. I will create an issue in github for this problem. Thank you for your comment
**Here is the result in master branch :**

bug master pan

**Comment by mirsky-work** [comment]
> I tried what you did in the master branch and found that there was the same issue. I will create an issue in github for this problem. Thank you for your comment

Thanks for verifying this and reporting #453!

**Comment by mirsky-work** [comment]
Sorry, closed by mistake (Monday morning...).

**Review by PatrickTasse — APPROVED** [review]

**Review by PatrickTasse — COMMENTED** [review]
The commit message mentions support for "+" and "-". This now will only come later.

The lines in commit message should wrap after 72 chars.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 30]
```diff
@@ -273,21 +283,123 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+            let zoomNumber = 1;
+            const position = this.getTimeX(this.positionXMove);
+            if (!iszoomIn) {
+                zoomNumber = -1;
+            }
+            const startPos = position - this.props.unitController.viewRange.start;
+            const endPos = this.props.unitController.viewRange.end - position;
+            newStartRange = position - (startPos - zoomNumber * (startPos * TEN_PERCENT));
+            newEndRange = position + (endPos - zoomNumber * (endPos * TEN_PERCENT));
+            this.updateRange(newStartRange, newEndRange);
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:395]: This creates a strange behavior in the following use case:
- Zoom to half the trace range
- Pan to beginning of trace
- Try to zoom out with 'S" while the mouse is very close to the right edge.

What happens is that the window range start is beyond the start time (< 0) and gets snapped back to 0 in time-graph-unit-controller. Then the effect is a very slow zoom out step.

What I would suggest is to calculate the new view range length based on zoom in/out (capped to absolute range), then apply this new window range centered on the mouse position. After this, snap the new window range to the bounds [0, absolute range] like it's done below for the pan case.

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/xy-output-component.tsx [hunk 31]
```diff
@@ -273,21 +283,125 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+            end: rangeEnd
+        };
+    }
+
+    private zoom(iszoomIn: boolean) {
+        if (this.props.unitController.viewRangeLength >= 1) {
+            let newStartRange = 0;
+            let newEndRange = 0;
+            const zoomNumber = iszoomIn ? 1 : -1;
+                const position = this.getTimeX(this.positionXMove);
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:386]: Indentation

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/xy-output-component.tsx [hunk 32]
```diff
@@ -273,21 +283,125 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+                const startPos = position - this.props.unitController.viewRange.start;
+                const endPos = this.props.unitController.viewRange.end - position;
+                newStartRange = position - (startPos - zoomNumber * (startPos * TEN_PERCENT));
+                newEndRange = position + (endPos - zoomNumber * (endPos * TEN_PERCENT));
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:390]: Seems like the end result of this code is that range is changed by a factor of 0.9 or 1.1.
But to be the mathematical inverse, the factor should be 0.9 or 1/0.9=1.11111111111.

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/xy-output-component.tsx [hunk 33]
```diff
@@ -114,10 +119,33 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+    componentWillUnmount(): void {
+        if (this.preventDefaultHandler !== undefined) {
+            const xyMainElement = document.getElementById(this.getUniqueDivId());
+            if (xyMainElement) {
+                xyMainElement.removeEventListener('wheel', this.preventDefaultHandler);
+            }
+            this.preventDefaultHandler = undefined;
+        }
+    }
+
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:161]: The timegraph is able to prevent the mouse wheel document zoom without resorting to adding/removing the handler and needing uniqueDivId. I think it would be worth to try and simplify the code by using a similar strategy.

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/xy-output-component.tsx (hunk 33 above)
Comment (Rodrigoplp-work) [reply packages/react-components/src/components/xy-output-component.tsx:161]: Patrick suggested using `React.createRef()` for this, and I agree. It could be implemented by:

1. Adding `private chartRef: any;` below line 41;

2. Adding `this.chartRef = React.createRef();` below line 91;

3. Replacing most of the code block above with `this.chartRef.current.addEventListener('wheel', this.preventDefaultHandler);`. No need for `xyMainElement`.

Also, in this case it seems that the eventListener is being cleared by the garbage collector. No need to remove it manually.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 34]
```diff
@@ -113,6 +120,14 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
         }
         if (this.lineChartRef.current) {
+            if (this.preventDefaultHandler === undefined) {
+                this.preventDefaultHandler = (event: WheelEvent) => {
+                    if (event.ctrlKey) {
+                        event.preventDefault();
+                    }
+                };
+                this.chartRef.current.addEventListener('wheel', this.preventDefaultHandler);
+            }
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:143]: Could the listener be added in componentDidMount instead?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 35]
```diff
@@ -165,12 +180,14 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
         // width={this.props.style.chartWidth}
         return <React.Fragment>
             {this.state.outputStatus === ResponseStatus.COMPLETED ?
-                <div id='xy-main' tabIndex={0}
+                <div id={this.getUniqueDivId()} tabIndex={0}
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:196]: I am not sure if this change is required anymore now that we have the chartRef. Could you check if it works without it?

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/xy-output-component.tsx (hunk 34 above)
Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx:143]: I tried it but it didn't work. the view didn't want to display.

File: packages/react-components/src/components/xy-output-component.tsx (hunk 35 above)
Comment (IbrahimFradj) [reply packages/react-components/src/components/xy-output-component.tsx]: i just tried now and it works.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 36]
```diff
@@ -394,31 +443,69 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+
+    private updateSelectionZoom() {
+        const xStartPos = this.state.zoomSelectionRangeStart;
+        this.setState({
+            zoomSelectionRangeStart: xStartPos,
+            zoomSelectionRangeEnd: this.getTimeX(this.positionXMove)
+        });
+    }
+
+    private resetSelectionZoom() {
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:475]: Since it actually does the zoom, I suggest calling it applySelectionZoom().

File: packages/react-components/src/components/xy-output-component.tsx [hunk 37]
```diff
@@ -394,31 +443,69 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
             }
-            case 'D':
-            case 'd':
-            case 'ArrowRight': {
-                this.pan(PAN_RIGHT);
-                break;
+        }
+    }
+
+    private updateSelectionZoom() {
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:467]: Since you already keep track of the mouse x position, you could use that to draw the selection range, then you only need to convert it to a time when applying the zoom selection.

This way you wouldn't need to update the zoom end time as you pan/zoom.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 38]
```diff
@@ -320,33 +355,42 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+                newStartRange = Math.max(newStartRange, 0);
+                newEndRange = Math.min(newEndRange, this.props.unitController.absoluteRange);
             }
-            newStartRange = Math.max(newStartRange, 0);
-            newEndRange = Math.min(newEndRange, this.props.unitController.absoluteRange);
             this.updateRange(newStartRange, newEndRange);
         }
     }
 
     private pan(panLeft: boolean) {
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:386]: Changes in this method might not be necessary if you just use the x position to draw the zoom selection.

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/xy-output-component.tsx [hunk 39]
```diff
@@ -394,31 +435,56 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
-        switch (key.key) {
-            case 'W':
-            case 'w': {
-                this.zoom(ZOOM_IN);
-                break;
+    private onMouseLeave(event: React.MouseEvent) {
+        this.isMouseLeave = true;
+        if (this.isRightClick) {
+            if (this.getTimeX(event.nativeEvent.offsetX) < this.props.unitController.viewRange.start) {
+                this.positionXMove = this.props.unitController.viewRange.start;
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:448]: Mixing timestamps with x-coordinates...

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

**Review by PatrickTasse — APPROVED** [review]

**Review by bhufmann — APPROVED** [review]
LGTM. Great improvement. Thank you very much!

**Review by PatrickTasse — APPROVED** [review]
